### DISCUSSION TITLE
Intermediate certificate not sent with Apple configuration profile

### DISCUSSION THREAD

**Discussion by hgl** [strongswan/strongswan#1609] [Q&A]
This isn't really a strongswan question, but I was following strongswan's doc on [generating p12 certificate](https://docs.strongswan.org/docs/5.9/pki/pkiQuickstart.html#_pkcs12_container) and [creating the Apple configuration profile](https://docs.strongswan.org/docs/5.9/interop/appleIkev2Profile.html#_certificate_based_authentication). Hope someone who has gone through this process could offer some help.

My initiator, which is an Apple device, couldn't establish a certificate-based IKEv2 connection with a StrongSwan responder. The issue was that the Apple device, only sent the end-entity certificate but not the intermediate certificate.

The end-entity certificate, its private key, and the intermediate certificate were bundled into pkcs12 with this command:
```
openssl pkcs12 -export \
	-inkey "$CLIENT_KEY" \
	-in "$CLIENT_CERT" \
	-certfile "$INTERMEDIATE_CERT" \
	-passout pass:0 \
	-legacy
```

It was then, along with the CA certificate, added to the Apple configuration profile. I can confirm they were all imported correctly into the Keychain. I can also confirm the chain is valid:

```
openssl verify -CAfile ca.crt -untrusted $INTERMEDIATE_CERT $CLIENT_CERT
```
returned ok.

The IKEv2 setting in the configuration profile is like this
```
<key>IKEv2</key>
<dict>
	<key>RemoteAddress</key>
	<string>example.com</string>
	<key>AuthenticationMethod</key>
	<string>Certificate</string>
	<key>CertificateType</key>
	<string>ECDSA256</string>
	<key>ServerCertificateIssuerCommonName</key>
	<string>CA CN</string>
	<key>RemoteIdentifier</key>
	<string>example.com</string>
	<key>LocalIdentifier</key>
	<string>localid</string>
	<key>PayloadCertificateUUID</key>
	<string>[uuid number]</string>
	<key>DeadPeerDetectionRate</key>
	<string>Low</string>
	<key>EnablePFS</key>
	<true/>
	<key>UseConfigurationAttributeInternalIPSubnet</key>
	<true/>
	<key>IKESecurityAssociationParameters</key>
	<dict>
		<key>DiffieHellmanGroup</key>
		<integer>19</integer>
		<key>EncryptionAlgorithm</key>
		<string>AES-128-GCM</string>
		<key>IntegrityAlgorithm</key>
		<string>SHA2-256</string>
		<key>LifeTimeInMinutes</key>
		<integer>60</integer>
	</dict>
	<key>ChildSecurityAssociationParameters</key>
	<dict>
		<key>DiffieHellmanGroup</key>
		<integer>19</integer>
		<key>EncryptionAlgorithm</key>
		<string>AES-128-GCM</string>
		<key>LifeTimeInMinutes</key>
		<integer>30</integer>
	</dict>
	<key>OnDemandEnabled</key>
	<false/>
</dict>
```

In charon's syslog, it contained `received end entity cert "CN=localid"` and `no issuer certificate found for "CN=localid"`.

I wonder what should I do to make the Apple device also send the intermediate certificate?

**Comment by tobiasbrunner** [comment]
> It was then, along with the CA certificate, added to the Apple configuration profile.

Does it make a difference if you add the intermediate CA certificate as a separate payload to the profile (using the `com.apple.security.pkcs1` or `com.apple.security.pem` type, although `com.apple.security.root` might work too as that seems to be just an alias for the former)?

> I can confirm they were all imported correctly into the Keychain.

So the intermediate CA certificate is actually extracted from the PKCS#12 container? At least for root CA certificates that didn't seem to work previously.

> I wonder what should I do to make the Apple device also send the intermediate certificate?

Is there any indication that Apple devices actually do that?

**Comment by hgl** [comment]
> Does it make a difference if you add the intermediate CA certificate as a separate payload to the profile

No.

> Is there any indication that Apple devices actually do that?

I don't think so, and after some more experimenting, I believe Apple devices actually aren't capable of sending intermediate certificates. I've redesigned my PKI to eliminate intermediate certificates for clients.

Thanks for the help.
